### ISSUE TITLE
tensorflow version of the DARE-GRAM loss

### ISSUE THREAD

**Issue by noushinha** [ismailnejjar/DARE-GRAM#2]
Hi,

Thanks a lot for the public repository of the DARE-GRAM paper.
I am using it for the DAR of an MLP model with 4 layers. 
Inputs are patches of `48x48x48`, and the output is a 6D vector of object pose.
The total loss is Huber loss plus Dare-GRAM loss. My problem right now is that the loss value increases for both training and validation.
Any idea what could possibly be wrong? I get very large values for `distval`, which is the scale term. Is this expected?
Would you please kindly check my implementation and let me know if it is correct?

Should I choose `alpha_cos` and `gamma_scale` coefficients empirically, or is there a better approach?

```


    def dare_gram(self, H1, H2):
        b, p = H1.shape

        e1 = tf.ones([b, 1], dtype=tf.float32)
        e2 = tf.ones([b, 1], dtype=tf.float32)
        A = tf.concat([e1, H1], axis=1)
        B = tf.concat([e2, H2], axis=1)

        cov_A = tf.matmul(tf.transpose(A), A)
        cov_B = tf.matmul(tf.transpose(B), B)

        L_A = tf.linalg.svd(cov_A, compute_uv=False)
        L_B = tf.linalg.svd(cov_B, compute_uv=False)

        eigen_A = tf.math.cumsum(tf.stop_gradient(L_A), axis=0) / tf.reduce_sum(L_A)
        eigen_B = tf.math.cumsum(tf.stop_gradient(L_B), axis=0) / tf.reduce_sum(L_B)

        if eigen_A[1] > 0.9:
            T = tf.stop_gradient(eigen_A[1])
        else:
            T = 0.9

        index_A = tf.where(tf.stop_gradient(eigen_A) <= T)[-1]

        if eigen_B[1] > 0.9:
            T = tf.stop_gradient(eigen_B[1])
        else:
            T = 0.9

        index_B = tf.where(tf.stop_gradient(eigen_B) <= T)[-1]

        if index_A[0] > index_B[0]:
            k = index_A[0]
        else:
            k = index_B[0]

        A = tf.linalg.pinv(cov_A, rcond=(L_A[k] / tf.stop_gradient(L_A[0])))
        B = tf.linalg.pinv(cov_B, rcond=(L_B[k] / tf.stop_gradient(L_B[0])))

        denominator = tf.maximum(tf.norm(A, ord=2, axis=(0)) * tf.norm(B, ord=2, axis=(0)), 1e-6)
        cosSim = tf.linalg.diag_part(tf.matmul(A, B) / denominator)
        identity = tf.ones([p + 1], dtype=tf.float32)
        pPlus1Constant = tf.constant(float(p + 1), dtype=tf.float32)
        cos = tf.math.divide(tf.norm(tf.subtract(identity, cosSim), ord=1, axis=(0)), pPlus1Constant)

        kConstant = tf.cast(k, dtype=tf.float32)
        distval = tf.math.divide(tf.norm(tf.subtract(tf.stop_gradient(L_A[:k]), tf.stop_gradient(L_B[:k])), ord=2), kConstant)

        return (0.1 * cos) + (0.0005 * distval)
```

**Comment by ismailnejjar** [comment]
Hey,
I noticed a potential issue in your implementation. In the last line calculating distval, you have tf.stop_gradient on both L_A and L_B:

`pythondistval = tf.math.divide(tf.norm(tf.subtract(tf.stop_gradient(L_A[:k]), tf.stop_gradient(L_B[:k])), ord=2), kConstant)`

This prevents gradients from flowing through the scale loss component. Please try removing the stop_gradient:

`pythondistval = tf.math.divide(tf.norm(tf.subtract(L_A[:k], L_B[:k]), ord=2), kConstant)`

Also, if you're still getting very large distval values after this fix, try adjusting your coefficient,  maybe start with alpha_cos=1.0 and gamma_scale=0.01 and tune from there based on the relative magnitudes of each loss component.

**Comment by noushinha** [comment]
Many thanks for the help.
I will try these changes in order and see if the loss starts to decrease.

**Comment by noushinha** [comment]
Unfortunately, even after these changes, the loss increases.
I believe there is a need for normalizing DARE-GRAM loss before adding to MSE loss because the ranges are not similar, and the gradients explode. Any idea how I can do that? Training only using the cosine similarity term decreases, so the problem is with the feature scaling part.

**Comment by noushinha** [comment]
I made the changes, but the total loss is still fluctuating as follows due to the numerical instability of the `feature_scale` term:

This plot shows the total loss, which includes `L_MSE + L_angle-scale + L_feature-scale`

<img width="657" height="526" alt="Image" src="https://github.com/user-attachments/assets/2b17af02-f41e-4372-a99d-15b6fbda7844" />

This plot is `L_MSE + L_angle-scale`

<img width="657" height="526" alt="Image" src="https://github.com/user-attachments/assets/7bc6323f-63e6-4faa-a4f1-2c5ddbcebbda" />

Also, SVD fails from time to time!
Any other idea what can possibly be wrong?
